### DISCUSSION TITLE
The manhattan router doesn't apply padding around elements and obstacles

### DISCUSSION THREAD

**Discussion by kun-acatise** [clientIO/joint#2738] [Q&A] [answered]
### Introduction

I am following the doc to set default router and anchor with this:
```
defaultRouter: {
    name: 'manhattan',
    args: {
        padding: 10
    }
},
defaultConnectionPoint: { name: 'anchor' },
defaultAnchor: function (elementView) {
        const link = this.model;
        const element = elementView.model;
        const source = link.getSourceElement();
        const target = link.getTargetElement();
        if (!source || !target) return element.getBBox().rightMiddle();
        const sourceStack = source.get('column');
        const targetStack = target.get('column');
        const sourceRow = source.get('row');
        const targetRow = target.get('row');
        if (sourceStack === targetStack) {
          if (source === element) {
            if (targetRow - sourceRow > 1) {
              return element.getBBox().rightMiddle();
            }
            return element.getBBox().bottomMiddle();
          } else {
            if (targetRow - sourceRow > 1) {
              return element.getBBox().rightMiddle();
            }
            return element.getBBox().topMiddle();
          }
        } else if (sourceStack > targetStack) {
          if (source === element) {
            return element.getBBox().leftMiddle();
          }
          return element.getBBox().rightMiddle();
        } else {
          if (source === element) {
            return element.getBBox().bottomMiddle();
          }
          console.log(element, source, element.getBBox().leftMiddle())
          return element.getBBox().leftMiddle();
        }
      },
```

But the result doesn't show the padding correct 
<img width="380" alt="image" src="https://github.com/user-attachments/assets/6335a378-d36b-40b6-a81a-42895872946d">

Does anybody has such similar experience? Thank you in advance. 

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment]
Is the background also a shape? If so, you would have to tell the `manhattan` router not to consider it an obstacle. See [excludeEnds](https://docs.jointjs.com/api/routers/#built-in-routers) option.

You can also add `fallbackRouter` option to verify if the route could not be found.
```ts
defaultRouter: {
    name: 'manhattan',
    args: {
        padding: 10,
        fallbackRouter: () => {
           debugger;
           return [];
        }
    }
},
```

**Reply by kun-acatise** [reply]
@kumilingus 
Thanks so much for your reply.
I removed all the other shapes and only keep the node and links but still the route looks strange.
<img width="638" alt="image" src="https://github.com/user-attachments/assets/bb790b96-766d-4455-b327-8e141accf44e">

And I tried with the `fallbackRouter`, it is not falling to the callback function which means the router can be found I guess. 
Please help if you have any clues thanks!

**Reply by kumilingus** [reply]
Could you please create a simple example (e.g. JSFiddle) that reproduces the issue? Thank you

**Reply by kun-acatise** [reply]
@kumilingus 
Thank you for replying! Here is the example. 
https://codesandbox.io/p/sandbox/serene-bartik-fps3qs

**Reply by kumilingus** [reply, marked as answer]
To control the direction of the `manhattan` router, the `startDirections` and `endDirections` options must be used.
Defining anchors at the element boundary is not sufficient in the case of `manhattan` (unlike the `rightAngle` router).
The router will examine all initial directions. In each direction, it will find the boundary of the shape in that direction and add extra distance equal to `padding`, and from there it will try to find a path using the A* algorithm.

You can rewrite your code using only the `defaultRouter` as shown below:

```ts
defaultRouter: function (vertices, _, linkView) {
  const link = linkView.model;
  const startDirections = [];
  const endDirections = [];
  if (link.hasLoop()) {
    startDirections.push("right");
    endDirections.push("right");
  } else {
    startDirections.push("bottom");
    endDirections.push("left");
  }
  return routers.manhattan(
    vertices,
    {
      startDirections,
      endDirections,
      padding: 20,
    },
    linkView
  );
},
```

**Reply by kun-acatise** [reply]
@kumilingus
wow, it works! Thank you so much for your quick reaction! That really helps!

**Reply by kun-acatise** [reply]
Hey @kumilingus 
I am trying to solve the issue you pointed out
> Is the background also a shape? If so, you would have to tell the manhattan router not to consider it an obstacle. See [excludeEnds](https://docs.jointjs.com/api/routers/#built-in-routers) option.

May I double check with you that it should be `excludeEnds` or `excludeTypes`?
The background type is `WorkflowDivider` so is this correct if I configure it like this ?
```
return routers.manhattan(
    vertices,
    {
       excludeTypes: [
              'WorkflowDivider',
      ],
      ...,
    },
    linkView
  );

```

Thank you!

**Reply by kumilingus** [reply]
Yeah, that looks good.

**Reply by alexandernst** [reply]
May I jump here and ask a related question? 😏

I have this https://codesandbox.io/p/sandbox/crimson-hill-nz8wwm , which is basically the same situation, but instead of anchors it's using ports. The thing is that there is one particular position in which the link stops honouring the start/end directions and it just goes "straight" towards the target port. That position is right where the source and the target are horizontally aligned.

![image](https://github.com/user-attachments/assets/27b8b6bf-cb91-4a77-aa15-5d41f18a875d)

Why is this happening?

**Reply by kumilingus** [reply]
Hi @alexandernst, I don't seem to have access to the example. Thanks!

**Reply by alexandernst** [reply]
@kumilingus sorry, my bad! The link should be accesible now 🙏

**Reply by alexandernst** [reply]
Hi @kumilingus ! Did you have a chance to look at this?

**Reply by kumilingus** [reply]
> What is the problem?

The `manhattan` router currently does not support ports positioned inside the elements.
It will find a start point outside the port, but the point could be still inside an obstacle i.e. inside the element body.

> Why does it work in all cases except when the ports are aligned?

it's just a coincidence, a rounding error. If you move the port further to the element's center, it will never find the route.

> Is there any workaround?

You can add 2 more vertices to help `manhattan` to find the route. Also, using the `rightAngle` router as a fallback router yields much better results.

```ts
defaultRouter: function(vertices, _, linkView) {
    const link = linkView.model;
    const startDirections = [];
    const endDirections = [];
    let firstVertex, lastVertex;
    if (link.hasLoop()) {
        startDirections.push('top');
        endDirections.push('top');
        firstVertex = linkView.sourceView.model.getBBox().bottomMiddle().offset(0, 30);
        lastVertex = linkView.targetView.model.getBBox().bottomMiddle().offset(0, 30);
    } else {
        startDirections.push('bottom');
        endDirections.push('top');
        firstVertex = linkView.sourceView.model.getBBox().bottomMiddle().offset(0, 30);
        lastVertex = linkView.targetView.model.getBBox().topMiddle().offset(0, -30);
    }
    return routers.manhattan(
        [firstVertex, ...vertices, lastVertex],
        {
            startDirections,
            endDirections,
            padding: 20,
            fallbackRouter: routers.rightAngle
        },
        linkView
    );
},
```

**Reply by alexandernst** [reply]
Thank you @kumilingus ! This is actually very helpful! 🙏
Let's keep the discussion in https://github.com/clientIO/joint/issues/2763 so we don't spam this discussion.
